confmap/provider/s3provider/provider.go

```diff
@@ -0,0 +1,153 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/provider/internal"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+const (
+	schemeName   = "s3"
+	stop         = "."
+	colonSlashes = "://"
+	s3URITail    = ".amazonaws.com/"
+)
+
+type provider struct{}
+
+// New returns a new confmap.Provider that reads the configuration from a file.
+//
+// This Provider supports "s3" scheme, and can be called with a "uri" that follows:
+//   s3-uri : s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
+//
+// One example for s3-uri be like: s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg
+//
+// Examples:
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{}
+}
+
+func (fmp *provider) Retrieve(_ context.Context, uri string, _ confmap.WatcherFunc) (confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+
+	// Check if users set up their env for S3 Auth check yet
+	if os.Getenv("AWS_ACCESS_KEY_ID") == "" || os.Getenv("AWS_SECRET_ACCESS_KEY") == "" {
+		return confmap.Retrieved{}, fmt.Errorf("unable to fetch access keys for S3 Auth")
+	}
+
+	// Split the uri and get [BUCKET], [REGION], [KEY]
+	bucket, region, key, err := S3URISplit(uri)
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not valid s3-url", uri)
+	}
+
+	// AWS SDK default config
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("AWS SDK's default configuration fail to load")
+	}
+
+	// s3 client / manager
+	client := s3.NewFromConfig(cfg)
+	downloader := manager.NewDownloader(client)
```

Comment:
Is this a downloader or a manager? Confused by the comment

Reply:
Hi Eric, sorry for the confusion.

I would like to change it as 

// to create a s3 client and also a s3 downloader
// s3 downloader is especially for s3 downloading op

Does it make sense?

Reply:
op is operation 
	
	// to create a s3 client and also a s3 downloader
	// s3 client provides interfaces for Bucket/File Management in Amazon S3
	// s3 downloader is especially for s3 downloading operation
	client := s3.NewFromConfig(cfg)
	downloader := manager.NewDownloader(client)

Reply:
@Aneurysm9 I thought I was misled by this [documentation](https://aws.github.io/aws-sdk-go-v2/docs/sdk-utilities/s3/), which uses download manager to do the downloading.

And thank you for clarifying the limitation of using download manager. I've pushed new codes and used client.GetObject() instead. Tested and was running well.